src/routes/campaigns.ts

```diff
@@ -235,3 +235,174 @@ export async function handleDeleteAllCampaigns(c: ContextWithAuth) {
     return c.json({ error: "Internal server error" }, 500);
   }
 }
+
+// Add resource to campaign
+export async function handleAddResourceToCampaign(c: ContextWithAuth) {
+  try {
+    const userAuth = (c as any).userAuth;
+    const campaignId = c.req.param("campaignId");
+    const { type, id, name } = await c.req.json();
+
+    console.log(
+      `[Server] POST /campaigns/${campaignId}/resource - starting request`
+    );
+    console.log("[Server] User auth from middleware:", userAuth);
+    console.log("[Server] Request body:", { type, id, name });
+
+    if (!type || !id) {
+      return c.json({ error: "Resource type and id are required" }, 400);
+    }
+
+    // First, check if the campaign exists and belongs to the user
+    const campaign = await c.env.DB.prepare(
+      "SELECT id, name, username FROM campaigns WHERE id = ? AND username = ?"
+    )
+      .bind(campaignId, userAuth.username)
+      .first<{ id: string; name: string; username: string }>();
+
+    if (!campaign) {
+      console.log(
+        `[Server] Campaign ${campaignId} not found or doesn't belong to user ${userAuth.username}`
+      );
+      return c.json({ error: "Campaign not found" }, 404);
+    }
+
+    console.log("[Server] Found campaign:", campaign);
+
+    // Check if resource already exists in this campaign
+    const existingResource = await c.env.DB.prepare(
+      "SELECT id, file_name FROM campaign_resources WHERE campaign_id = ? AND file_key = ?"
+    )
+      .bind(campaignId, id)
+      .first<{ id: string; file_name: string }>();
+
+    if (existingResource) {
+      console.log(
+        `[Server] Resource ${id} already exists in campaign ${campaignId}`
+      );
+      // Return success instead of error - this is idempotent behavior
+      return c.json(
+        {
+          resource: {
+            id: existingResource.id,
+            campaignId,
+            fileKey: id,
+            fileName: existingResource.file_name,
+            description: "",
+            tags: "[]",
+            status: "active",
+            createdAt: new Date().toISOString(),
+            updatedAt: new Date().toISOString(),
+          },
+          message: "Resource already exists in this campaign",
+        },
+        200
+      );
+    }
+
+    // Add the resource to the campaign
+    const resourceId = crypto.randomUUID();
+    const now = new Date().toISOString();
+
+    await c.env.DB.prepare(
+      "INSERT INTO campaign_resources (id, campaign_id, file_key, file_name, description, tags, status, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)"
+    )
+      .bind(
+        resourceId,
+        campaignId,
+        id,
+        name || id,
+        "",
+        "[]",
+        "active",
+        now,
+        now
+      )
+      .run();
+
```

Comment:
at some point around here we need to kick off indexation into the rag db for the campaign. not suggesting do that now, just worth a note (maybe a TODO?)